ons-spark/spark-concepts/exercises.md

```diff
@@ -0,0 +1,233 @@
+rescue_with_pop.count()
+
+# Get count of each animal group
+animal_count = (rescue
+                .groupBy("animal_group")
+                .agg(F.count("incident_number").alias("animal_count")))
+
+# Join this back to the rescue_with_pop DF
+rescue_with_pop = rescue_with_pop.join(animal_count, on="animal_group", how="left")
+
+# Filter out animals with small counts
+rescue_with_pop = rescue_with_pop.filter(F.col("animal_count") >= 5)
+
+# Sort the final data
+rescue_with_pop = rescue_with_pop.orderBy(F.desc("total_cost"))
+
+# Get the top 10 and row count
+top_10 = rescue_with_pop.limit(10).toPandas()
+row_count = rescue_with_pop.count()
+```
+
+```{code-tab} r R
+library(sparklyr)
+library(dplyr)
+
+config <- yaml::yaml.load_file("ons-spark/config.yaml")
+
+exercise_config <- sparklyr::spark_config()
+# Disable broadcast join by default
+exercise_config$spark.sql.autoBroadcastJoinThreshold <- -1
+
+sc <- sparklyr::spark_connect(
+    master = "local[2]",
+    app_name = "exercises",
+    config = exercise_config)
+
+# Read source data
+rescue <- sparklyr::spark_read_parquet(sc, config$rescue_path) %>%
+    sparklyr::select(incident_number, animal_group, total_cost, postcode_district)
+population <- sparklyr::spark_read_parquet(sc, config$population_path)
+
+
+# Sort and preview source data
+rescue <- rescue %>% sparklyr::sdf_sort("incident_number")
+rescue %>% head(5) %>% sparklyr::collect()
+rescue %>% sparklyr::sdf_nrow()
+population <- population %>% sparklyr::sdf_sort("postcode_district")
+population %>% head(5) %>% sparklyr::collect()
+population %>% sparklyr::sdf_nrow()
+
+# Join population data to rescue data
+rescue_with_pop <- rescue %>%
+    sparklyr::left_join(population, by="postcode_district")
+
+# Sort and preview
+rescue_with_pop <- rescue_with_pop %>% sparklyr::sdf_sort("incident_number")
+rescue_with_pop %>% head(5) %>% sparklyr::collect()
+rescue_with_pop %>% sparklyr::sdf_nrow()
+
+# Get count of each animal group
+animal_count <- rescue_with_pop %>%
+    dplyr::group_by(animal_group) %>%
+    dplyr::summarise(animal_count = n())
+
+# Join this back to the rescue_with_pop DF
+rescue_with_pop <- rescue_with_pop %>%
+    sparklyr::left_join(animal_count, by="animal_group")
+
+# Filter out animals with small counts
+rescue_with_pop <- rescue_with_pop %>%
+    sparklyr::filter(animal_count >= 5)
+
+# Sort the final data
+rescue_with_pop <- rescue_with_pop %>%
+    dplyr::arrange(desc(total_cost))
+
+# Get the top 10 and row count
+top_10 <- rescue_with_pop %>% head(10) %>% sparklyr::collect()
+row_count <- rescue_with_pop %>% sparklyr::sdf_nrow()
+```
+````
+<details>
+<summary>Answer</summary>
+
+Using the ideas from the Optimising and Avoiding Shuffle chapter, we can:
+- Minimise actions: many of the `.show()`/`head(n) %>% collect` and `.count()`/`sdf_nrow()` actions are not needed. Just keep the ones at the end.
+- Caching: The two actions at the end are needed, so to avoid full recalculation of the DF, use a cache.
```

Comment:
I'm guessing this would only be a small gain here because there is only one action being called after the cache. But the principle if most definitely valid.